### PR TITLE
Added ICON interpolator (to point locations)

### PR DESCRIPTION
**Pull request by efmkoene**
This branch adds a tool to the `tools` folder, which can interpolate ICON output to point locations, using inverse distance weighted interpolation horizontally, and linear interpolation vertically. [*]

```python
import sys
sys.path.insert(1, '/path/to/processing-chain/tools')
from ICON_to_point import ICON_to_point

longitudes = [15,13,14,15] # Self-explanatory: the longitude coordinate of the station
latitudes = [51,52,50,51] # Self-explanatory: the latitude coordinate of the station
asl = [0,1000,1200,0] # The true topographical height (above sea level, asl) of the station (only needed for mountain stations)
elev = [10,70,400,40] # The inlet height above the surface
strategy = ['ground', 'mountain', 'mountain', 'middle'] # Ground stations: model topography[i] + elev[i]; Mountain stations: asl[i] + elev[i]; middle: half ground and half mountain station approach

# Example 3: get two fields
ICON_to_point(longitude=longitudes, latitude=latitudes, station_ground_elevation=asl,
                            station_altitude_over_ground=elev,                            
                            ICON_field_path='/scratch/snx3000/ekoene/processing-chain/work/CTDAS_OCO2/2018010100_0_8664/icon/output_2018010100_opt/ICON-ART-UNSTR_2018-01-02T08:00:00.000.nc', 
                            ICON_grid_path='/scratch/snx3000/ekoene/icon_europe_DOM01.nc', 
                            interpolation_strategy=strategy, 
                            k = 4, field_name=['qv','temp'])
```

It can also be called from the command line (but with reduced performance, as only one lon/lat point can be interpolated at a time).

It relies on a relatively modern `xarray` version (I have `2023.7.0`), but that is automatically the case with the way the current conda environment is set up anyhow.

From the review, I would mainly ask for a good descriptive word that makes the difference between `station_ground_elevation` and `station_altitude_over_ground` as clear as possible -- it probably isn't a 100% clear right now. This is what it should correspond to:
* for the Cabauw station in the Netherlands, we have that `station_ground_elevation=0` (as that is the local topography) while `station_altitude_over_ground=27` (as that is the inlet height over the ground)
* for the Jungfraujoch station in Switzerland, we have that `station_ground_elevation=3580` (as that is the local topography) while `station_altitude_over_ground=5` (as that is the inlet height over the ground)

[*] _(As a curious sidenote, inverse distance weighted interpolation == linear interpolation when only 2 points are considered. So, really, it is using inverse distance weighing both horizontally and vertically)._

### CODE REVIEW

**Comment by mjaehn** [comment]
@efmkoene Thanks for this PR! I will review this one and already set the base branch to v3.1-rc. Since here, just one file is added, there is no conflict.

File: jobs/tools/ICON_to_point.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+# TODO: re-use the interpolator & sampler when the function is called with a list of files.
+#       I've tested some options, but I fail to achieve better than linear scaling of the
+#       performance, whereas the second (etc.) calls to the function can entirely skip
+#       recomputing the interpolation weights etc. Probably just a limitation of this method
+#       and a limitation of passing around and merging/concatenating xarray data.
+
+
+def get_horizontal_distances(longitude, latitude, ICON_grid_path, k=5):
+    '''
```
Comment (mjaehn) [inline jobs/tools/ICON_to_point.py:15]: For consistency, always use """triple double quotes""" around docstrings.

File: jobs/tools/ICON_to_point.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        0] - 1  # If no result found: sample lies below lowest model level. Set it to the lowest model level
+
+    # Sample the corresponding vertical distances between the target and the model levels
+    vertical_distances = np.take_along_axis(
+        (model_levels - target_altitude).values, vertical_indices, axis=0)
+
+    return np.abs(vertical_distances).T, vertical_indices.T
+
+
+def ICON_to_point(longitude,
```
Comment (mjaehn) [inline jobs/tools/ICON_to_point.py:153]: Please use lowercase for function names. I would also suggest du replace `ICON` with `icon` in variable names.
```suggestion
def icon_to_point(longitude,
```

File: jobs/tools/ICON_to_point.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+import xarray as xr
+from sklearn.neighbors import BallTree
+from scipy import argmin
+import argparse
+
+# TODO: re-use the interpolator & sampler when the function is called with a list of files.
+#       I've tested some options, but I fail to achieve better than linear scaling of the
+#       performance, whereas the second (etc.) calls to the function can entirely skip
+#       recomputing the interpolation weights etc. Probably just a limitation of this method
+#       and a limitation of passing around and merging/concatenating xarray data.
```
Comment (mjaehn) [inline jobs/tools/ICON_to_point.py:11]: Is this note needed? If yes, please add your name/initials.

File: jobs/tools/ICON_to_point.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    # NB: the 'distances' are in units of radians; i.e., it assumes the Earth is a unit sphere!
+    # To get realistic distances, you need to multiply 'distances' with 6371e3 meters, i.e., the
+    # radius of the earth. However, such a constant factor cancels out when we compute the
+    # horizontal interpolation weights (which are normalized!), so there is no need to apply the
+    # multiplication with 6371e3.
+
+    return distances, indices
+
+
+def get_nearestvertical_distances(model_topography, model_levels,
```
Comment (mjaehn) [inline jobs/tools/ICON_to_point.py:80]:
```suggestion
def get_nearest_vertical_distances(model_topography, model_levels,
```

File: jobs/tools/ICON_to_point.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+                                  station_ground_elevation,
```
Comment (mjaehn) [inline jobs/tools/ICON_to_point.py:81]:
```suggestion
                                  station_height_msl,
```

File: jobs/tools/ICON_to_point.py [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+                                  station_altitude_over_ground,
```
Comment (mjaehn) [inline jobs/tools/ICON_to_point.py:82]:
```suggestion
                                  station_height_agl,
```

**Review by mjaehn — CHANGES_REQUESTED** [review]
Some minor comments

File: jobs/tools/ICON_to_point.py (hunk 3 above)
Comment (efmkoene) [reply jobs/tools/ICON_to_point.py:11]: OK!

File: jobs/tools/ICON_to_point.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
+import numpy as np
+import xarray as xr
+from sklearn.neighbors import BallTree
```
Comment (mjaehn) [inline jobs/tools/ICON_to_point.py:3]: `sklearn.neighbors` is not part of the conda environment. It should be added to `env/environment.yml`.

Comment (efmkoene) [reply jobs/tools/ICON_to_point.py:3]: Done. Thanks!

File: jobs/tools/ICON_to_point.py (hunk 2 above)
Comment (efmkoene) [reply jobs/tools/ICON_to_point.py]: OK, done!

**Comment by efmkoene** [comment]
Hopefully it's up to scratch now!

**Comment by mjaehn** [comment]
launch jenkins

**Review by mjaehn — APPROVED** [review]

**Pull request merged**